janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java

```diff
@@ -490,5 +409,189 @@ private static boolean coversAll(final MixedIndexType index, Condition<JanusGrap
+                && indexCoversOrder((MixedIndexType) bestIndexes.get(0).getIndex(), orders));
+            for (IndexCandidate c : bestIndexes) {
+                addToJointQuery(c, jointQuery);
+            }
+        }
+        return jointQuery;
+    }
 
+    /**
+     * Iterate over all potential indexes and compute a score based on how many clauses
+     * this index covers. The index with the highest score (as long as it covers at least one additional clause)
+     * is picked and added to the joint query for as long as such exist.
+     * @param rawCandidates
+     * @param conditions
+     * @param coveredClauses
+     * @param isSorted
+     * @return
+     */
+    private JointIndexQuery selectIndicesByApprox(final Set<IndexType> rawCandidates, final MultiCondition<JanusGraphElement> conditions,
+                                                  final Set<Condition> coveredClauses, final MutableBoolean isSorted) {
+        final JointIndexQuery jointQuery = new JointIndexQuery();
+        while (true) {
+            IndexCandidate bestCandidate = null;
+            boolean candidateSupportsSort = false;
+
+            for (final IndexType index : rawCandidates) {
+                final IndexCandidate indexCandidate = enrichIndex(index, conditions);
+                if (indexCandidate == null) continue;
+
+                // start calculating score
+                double score = 0.0;
+                boolean coversAdditionalClause = false;
+                for (final Condition c : indexCandidate.getSubcover()) {
+                    double s = getConditionBasicScore(c);
+                    if (coveredClauses.contains(c)) s=s*ALREADY_MATCHED_ADJUSTOR;
+                    else coversAdditionalClause = true;
+                    score += s + getIndexTypeScore(index);
+                }
+                boolean supportsSort = orders.isEmpty() ||
+                    coveredClauses.isEmpty() && index.isMixedIndex() && indexCoversOrder((MixedIndexType) index, orders);
+                if (supportsSort) score+=ORDER_MATCH;
+                indexCandidate.setScore(score);
+                if (coversAdditionalClause && (bestCandidate == null || score > bestCandidate.getScore())) {
+                    bestCandidate = indexCandidate;
+                    candidateSupportsSort = supportsSort;
+                }
+            }
+            if (bestCandidate!=null) {
+                if (coveredClauses.isEmpty()) isSorted.setValue(candidateSupportsSort);
+                coveredClauses.addAll(bestCandidate.getSubcover());
+                addToJointQuery(bestCandidate, jointQuery);
+            } else {
+                break;
+            }
+        }
+        return jointQuery;
+    }
+
+    /**
+     * Validate given index candidate and enrich it
+     * @param index
+     * @param conditions
+     * @return
+     */
+    private IndexCandidate enrichIndex(final IndexType index, final MultiCondition<JanusGraphElement> conditions) {
```

Comment:
It looks like the only purpose of this method is to construct an instance of `IndexCandidate` from an `IndexType` and a `MultiCondition`, am I right?
If so, it would make sense to use this as a constructor of `IndexCandidate`.